# Python project

The information given was the column titles and questions to be answered.

Column titles and descriptions:

1. Category: The category to which this IMU belongs. 
2. Manufacturer: Name of the manufacturer. 
3. part_number: Discrete part number 
4. sku: stock keeping unit 
5. alt_sku: alternate stock keeping unit 
6. Description: description of the product 
7. Status: Whether it is instock or not 
8. Quantity: Quantity available 
9. Price: Price of the product 
10. sku_create_date: when was the product procured.

Below are the questions to be answered for this project:

For data In Stock: 

1. Which are the top Manufacturers with price, quantity?
2. Which are the most profitable products? 
3. Which is the product where you have seen a decrease in the price? 
4. Which is the product where you have seen an increase in the price? 
5. Which is the top category by price and quantity? 
6. Which products are aging? 
7. How to Decrease Product aging? 

For Data Not in Stock 

8. Which are the top Manufacturers with price, quantity? 
9. Which are the most profitable products? 
10. Which is the product where you have seen a decrease in the price? 
11. Which is the product where you have seen an increase in the price? 
12. Which is the top category by price and quantity?



**Firstly we must import dataset, then analyse and clean it.**

In [1]:
import pandas as pd
df = pd.concat(pd.read_excel(r'C:\Users\melis\OneDrive\Documents\imus.xlsx', sheet_name = None), ignore_index=True)#required to read all sheets
print(df)

         category                manufacturer part_number     sku alt_sku  \
0            1539  MS-OPEN VALUE SUBSCRIPTION   021-09108  BA0011     NaN   
1            1539  MS-OPEN VALUE SUBSCRIPTION   059-07014  BA0025     NaN   
2            1539  MS-OPEN VALUE SUBSCRIPTION   065-06444  BA0039     NaN   
3            1539  MS-OPEN VALUE SUBSCRIPTION   076-04355  BA0067     NaN   
4            1582  MS-OPEN VALUE SUBSCRIPTION   077-05318  BA0080     NaN   
...           ...                         ...         ...     ...     ...   
2627921      9945        VERBATIM CORPORATION       93975  986034     NaN   
2627922      9945        VERBATIM CORPORATION       93975  986034     NaN   
2627923      9945        VERBATIM CORPORATION       93975  986034     NaN   
2627924      9945        VERBATIM CORPORATION       93975  986034     NaN   
2627925      9945        VERBATIM CORPORATION       93975  986034     NaN   

                                               description    status  qty  

In [2]:
df.head(10)

category                manufacturer part_number     sku alt_sku  \
0      1539  MS-OPEN VALUE SUBSCRIPTION   021-09108  BA0011     NaN   
1      1539  MS-OPEN VALUE SUBSCRIPTION   059-07014  BA0025     NaN   
2      1539  MS-OPEN VALUE SUBSCRIPTION   065-06444  BA0039     NaN   
3      1539  MS-OPEN VALUE SUBSCRIPTION   076-04355  BA0067     NaN   
4      1582  MS-OPEN VALUE SUBSCRIPTION   077-05318  BA0080     NaN   
5      1539  MS-OPEN VALUE SUBSCRIPTION   079-04668  BA0095     NaN   
6      1539  MS-OPEN VALUE SUBSCRIPTION   125-00511  BA0224     NaN   
7      1539  MS-OPEN VALUE SUBSCRIPTION   126-01023  BA0249     NaN   
8      1539  MS-OPEN VALUE SUBSCRIPTION   126-01024  BA0250     NaN   
9      1539  MS-OPEN VALUE SUBSCRIPTION   164-05581  BA0291     NaN   

                                         description    status  qty   price  \
0               1YR OFFICESTD ALNG LICSAPK OLV, D AP  No Stock    0  143.58   
1                    1YR WORD ALNG LICSAPK OLV D AP,  No Stock    0   56.09   
2                   1YR EXCEL ALNG LICSAPK OLV D AP,  No Stock    0   56.09   
3                   1YR PRJCT ALNG LICSAPK OLV D AP,  No Stock    0  217.60   
4                  1YR ACCESS ALNG LICSAPK OLV D, AP  No Stock    0   56.09   
5                1YR PWRPOINT ALNG LICSAPK OLV D, AP  No Stock    0   56.09   
6          1YR VSTEAMFNDTNSVR ALNG LICSAPK, OLV D AP  No Stock    0  125.15   
7  1YR VSTEAMFNDTNSVRCAL ALNG, LICSAPK OLV D AP D...  No Stock    0  124.76   
8  1YR VSTEAMFNDTNSVRCAL ALNG, LICSAPK OLV D AP U...  No Stock    0  143.58   
9                  1YR PBLSHR ALNG LICSAPK OLV D, AP  No Stock    0   56.09   

  sku_create_date  
0      2020-08-17  
1      2020-08-17  
2      2020-08-17  
3      2020-08-17  
4      2020-08-17  
5      2020-08-17  
6      2020-08-17  
7      2020-08-17  
8      2020-08-17  
9      2020-08-17

In [3]:
df.shape #gives number of rows and columns

(2627926, 10)

In [4]:
df.info()#look into data types for each column

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2627926 entries, 0 to 2627925
Data columns (total 10 columns):
 #   Column           Dtype         
---  ------           -----         
 0   category         int64         
 1   manufacturer     object        
 2   part_number      object        
 3   sku              object        
 4   alt_sku          object        
 5   description      object        
 6   status           object        
 7   qty              int64         
 8   price            float64       
 9   sku_create_date  datetime64[ns]
dtypes: datetime64[ns](1), float64(1), int64(2), object(6)
memory usage: 200.5+ MB


In [5]:
missingvalues=df.isnull().sum()#check total missing values for each column on the data frame

In [6]:
missingvalues/ len(df)#percentage of values missing

category           0.000000
manufacturer       0.000000
part_number        0.000000
sku                0.000000
alt_sku            0.994488
description        0.000000
status             0.000000
qty                0.000000
price              0.000000
sku_create_date    0.000000
dtype: float64

In [7]:
 df.nunique()

category              153
manufacturer         1484
part_number        334574
sku                379112
alt_sku              2079
description        339222
status                  2
qty                  2618
price              122688
sku_create_date        10
dtype: int64

In [8]:
percentageduplicates=(len(df) - df.nunique())/len(df)#duplicates per column
print(percentageduplicates)

category           0.999942
manufacturer       0.999435
part_number        0.872685
sku                0.855737
alt_sku            0.999209
description        0.870916
status             0.999999
qty                0.999004
price              0.953314
sku_create_date    0.999996
dtype: float64


As can be seen above, column alt_sku is missing 99.6% of the values and therefore it must be removed. We can also see that all columns have above 90% duplicates. However, this is due to the nature of the data. Therefore, at this moment, the duplicates have not been removed. 
It can aso be seen above tha there are 153 categories and 1484 manufacturers.

In [9]:
df=df.drop(['alt_sku'], axis=1)#this column has too many missing values

In [10]:
df = df.drop_duplicates()

In [11]:
df.shape

(1105970, 9)

In [12]:
df["revenue"] = df["price"] * df["qty"]#need to create new variable for profitability which is revenue-cost of product.
#We do not have cost of product and therefore only revenue can be used.

In [13]:
instockdf=df.loc[df['status'] == 'In Stock']#filter to get values in stock
print(instockdf.head(10))
instockdf.shape

     category                    manufacturer         part_number     sku  \
253      9920  CASE LOGIC-PERSONAL & PORTABLE             3200943  BA6371   
271       531                OMNITRON SYSTEMS              7206-0  BA6834   
320      9070         ZEBRA ENTERPRISE MCD-A1       SAC5070-800CR  BC0227   
324      9080         ZEBRA ENTERPRISE MCD-A1  KT-CLMPT-RS507-01R  BC0232   
326      9080         ZEBRA ENTERPRISE MCD-A1    KT-PAD-RS507-10R  BC0234   
327      9080         ZEBRA ENTERPRISE MCD-A1  KT-STRPT-RS507-10R  BC0235   
328      9080         ZEBRA ENTERPRISE MCD-A1    KT-BKL-RS507-10R  BC0237   
330      9045                       MAGTEK NR            22523009  BC0283   
335      9070         ZEBRA ENTERPRISE MCD-A1           WA3017-G2  BC0389   
339      9070         ZEBRA ENTERPRISE MCD-A1        11-08062-02R  BC0437   

                                           description    status  qty   price  \
253  LAPTOP ROLLING CASE BLACK, SECURITY FRIENDLY 17IN  In Stock    7  

(156814, 10)

In [14]:
nostockdf=df.loc[df['status'] == 'No Stock']#filter to get values not in stock
print(nostockdf.head(10))
nostockdf.shape

   category                manufacturer part_number     sku  \
0      1539  MS-OPEN VALUE SUBSCRIPTION   021-09108  BA0011   
1      1539  MS-OPEN VALUE SUBSCRIPTION   059-07014  BA0025   
2      1539  MS-OPEN VALUE SUBSCRIPTION   065-06444  BA0039   
3      1539  MS-OPEN VALUE SUBSCRIPTION   076-04355  BA0067   
4      1582  MS-OPEN VALUE SUBSCRIPTION   077-05318  BA0080   
5      1539  MS-OPEN VALUE SUBSCRIPTION   079-04668  BA0095   
6      1539  MS-OPEN VALUE SUBSCRIPTION   125-00511  BA0224   
7      1539  MS-OPEN VALUE SUBSCRIPTION   126-01023  BA0249   
8      1539  MS-OPEN VALUE SUBSCRIPTION   126-01024  BA0250   
9      1539  MS-OPEN VALUE SUBSCRIPTION   164-05581  BA0291   

                                         description    status  qty   price  \
0               1YR OFFICESTD ALNG LICSAPK OLV, D AP  No Stock    0  143.58   
1                    1YR WORD ALNG LICSAPK OLV D AP,  No Stock    0   56.09   
2                   1YR EXCEL ALNG LICSAPK OLV D AP,  No Stock    0  

(949156, 10)

Now we can begin to answer the problem questions.

## Analysis of products in stock

In [44]:
#Which are the top Manufacturers with price, quantity?
print("TOP 5 MANUFACTURERS WITH THE HIGHEST PRICE")
print(instockdf.sort_values(by =['price'] , ascending=False)['manufacturer'].drop_duplicates().head(5))
print("****" * 30)

print("TOP 5 MANUFACTURERS PROVIDING HIGHEST QUANTITY")
print(instockdf.sort_values(by =['qty'] , ascending=False)['manufacturer'].drop_duplicates().head(5))
print("****" * 30)

#Which are the most profitable products?
print("TOP 5 MOST PROFITABLE PRODUCT")
print(instockdf.groupby(['manufacturer'])['revenue'].sum().sort_values(ascending=False).head(5))
print("****" * 30)
#create a new column which shows the difference in price from day to day
instockdf['price_difference'] = instockdf.sort_values('sku_create_date').groupby('description')['price'].diff().fillna(0) 

#Which is the product where you have seen a decrease in the price?
print("PRODUCT WITH PRICE DECREASE")
price_decrease=instockdf[instockdf['price_difference'] < 0.0].sort_values(by ='price_difference').head(1)
print(price_decrease['description'])
print("****" * 30)

#Which is the product where you have seen an increase in the price?
print("PRODUCT WITH PRICE INCREASE")
price_increase=instockdf[instockdf['price_difference'] > 0.0].sort_values(by ='price_difference',ascending=False).head(1)
print(price_increase['description'])
print("****" * 30)

#Which is the top category by price and quantity
print("TOP CATEGORY FOR PRICE", instockdf.sort_values(by = 'price', ascending=False).groupby("category").head(1)["category"].values[0])

print("****" * 30)

print("TOP CATEGORY FOR QUANTITY", instockdf.sort_values(by = 'qty', ascending=False).groupby("category").head(1)["category"].values[0])
print("****" * 30)



TOP 5 MANUFACTURERS WITH THE HIGHEST PRICE
1199874          HPE - SWITCHING
2620934         HPE - NIMBLE BTO
1203979      HPE - ARUBA INSTANT
1127383             PULSE SECURE
1017049    MELLANOX TECHNOLOGIES
Name: manufacturer, dtype: object
************************************************************************************************************************
TOP 5 MANUFACTURERS PROVIDING HIGHEST QUANTITY
2364017    DELL CSG SERVICE WARRANTIES
668949                             TMS
1698155                       CONFIG 6
2477656            DELL ESG WARRANTIES
785899                  MICROSOFT- ESD
Name: manufacturer, dtype: object
************************************************************************************************************************
TOP 5 MOST PROFITABLE PRODUCT
manufacturer
DELL CSG SERVICE WARRANTIES      2.308292e+09
DELL ESG WARRANTIES              2.010732e+09
HP INC. - SB NOTEBOOKS           2.492580e+08
MICROSOFT- ESD                   1.588981e+08
HP INC. - SB 

<ipython-input-44-397809cb237e>:15: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  instockdf['price_difference'] = instockdf.sort_values('sku_create_date').groupby('description')['price'].diff().fillna(0)


Need to determine which products have been there the longest therefore must subtract date in from the last date in the data.

In [35]:
#Which product are aging?
print("PRODUCT AGING")
df_in_stock = df[df["status"] == "In Stock"]

products_in_stock = list(set(df_in_stock['description']))
products_aged=pd.DataFrame(columns = ["Product", "Age"])

for i in products_in_stock:
    first_col = df_in_stock[df_in_stock['description'] == i]
    date1 = first_col.sku_create_date.min()
    date2 = first_col.sku_create_date.max()
    age = (date2-date1).days
    products_aged = products_aged.append({"Product" : i, "Age" : age}, ignore_index = True)


PRODUCT AGING


In [36]:
aging_products = products_aged[products_aged['Age']>7]
print("Total number of aged products: ", len(aging_products["Product"]))
print("Total number of products: ", len(products_in_stock))

print("Percentage of aged products", round((len(aging_products["Product"]) * 100) / len(products_aged),2), "%"  )

Total number of aged products:  13144
Total number of products:  30574
Percentage of aged products 42.99 %


In [39]:
print(aging_products.sort_values(by='Age',ascending= False).head(10))

                                                 Product Age
15198  ENS202EXT 11BGN 300MB 2.4GHZ, WPA WPA2 DETACHA...  11
15685        RJ45 10GBPS SFP+ TRANS MODULE, SFP+ TO RJ45  11
15829  6FT 3-SLOT AC POWER REPLACEMENT, F NOTEBOOKS I...  11
15820                               LIVE GAMER 4K GC573,  11
15817  3FT CAT6 PATCH CABLE SLIM BLUE, GIGABIT MOLDED...  11
15815  78KEY SLIM MINI USB KEYBOARD, INTEGRATED MULTI...  11
15810                     BLACKWIRE C5210 USB-A BULK-WW,  11
15807  DELL SNPTN78YC/32G COMP RDIMM, 32G DDR4-2666MH...  11
15805  WASP BARCODE CCD SCANNER FOR PC, WCS3900 W/6FT...  11
15803  4TB USB 3.0/3.1 GEN 1 7200 RPM, EXT HDD ALUMIN...  11


In order to reduce aging the company can:
- Sales or offers 
- Upselling 
- liquidate inventory
- donate aged stock

## Analysis for products not in stock

In [47]:
#Which are the top Manufacturers with price, quantity?
print("TOP 5 MANUFACTURERS WITH THE HIGHEST PRICE")
print(nostockdf.sort_values(by =['price'] , ascending=False)["manufacturer"].drop_duplicates().head(5))
print("****" * 30)

print("TOP 5 MANUFACTURERS PROVIDING HIGHEST QUANTITY")
print(nostockdf.sort_values(by =['qty'] , ascending=False)["manufacturer"].drop_duplicates().head(5))
print("****" * 30)

#Which are the most profitable products?
print("TOP 5 MOST PROFITABLE PRODUCT")
print(nostockdf.groupby(['manufacturer'])['revenue'].sum().sort_values(ascending=False).head(5))
print("****" * 30)

#create a new column which shows the difference in price from day to day
nostockdf['price_difference'] = nostockdf.sort_values('sku_create_date').groupby('description')['price'].diff().fillna(0) 

#Which is the product where you have seen a decrease in the price?
print("PRODUCT WITH PRICE DECREASE")
price_decrease=nostockdf[nostockdf['price_difference'] < 0.0].sort_values(by ='price_difference').head(1)
print(price_decrease['description'])
print("****" * 30)

#Which is the product where you have seen an increase in the price?
print("PRODUCT WITH PRICE INCREASE")
price_increase=nostockdf[nostockdf['price_difference'] > 0.0].sort_values(by ='price_difference',ascending=False).head(1)
print(price_increase['description'])
print("****" * 30)

#Which is the top category by price and quantity
print("TOP CATEGORY FOR PRICE", nostockdf.sort_values(by = 'price', ascending=False).groupby("category").head(1)["category"].values[0])

print("****" * 30)

print("TOP CATEGORY FOR QUANTITY", nostockdf.sort_values(by = 'qty', ascending=False).groupby("category").head(1)["category"].values[0])
print("****" * 30)

TOP 5 MANUFACTURERS WITH THE HIGHEST PRICE
1046052                         ALTNET
973610            JUNIPER ROW SERVICES
997704      JUNIPER ENTERPRISE ROUTING
1653726         CISCO - CCW MULTI YEAR
971703     HPE - ENTERPRISE NETWORKING
Name: manufacturer, dtype: object
************************************************************************************************************************
TOP 5 MANUFACTURERS PROVIDING HIGHEST QUANTITY
0          MS-OPEN VALUE SUBSCRIPTION
1757648                    PLANAR LFD
1757650     SUPERMICRO - MOTHERBOARDS
1757651                       PANDUIT
1757655         BLACK BOX CORPORATION
Name: manufacturer, dtype: object
************************************************************************************************************************
TOP 5 MOST PROFITABLE PRODUCT
manufacturer
22MILES                          0.0
PHOENIX AUDIO TECHNOLOGIES       0.0
PANASONIC PANABOARDS             0.0
PANASONIC FLAT PANEL DISPLAYS    0.0
PANASONIC CTO           

<ipython-input-47-dc7780b83a9f>:16: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  nostockdf['price_difference'] = nostockdf.sort_values('sku_create_date').groupby('description')['price'].diff().fillna(0)


PRODUCT WITH PRICE DECREASE
1653733    CCW ONLY SWSS UPG 3YR DISTI SO=, CSC IA CLD 3....
Name: description, dtype: object
************************************************************************************************************************
PRODUCT WITH PRICE INCREASE
388652    CCW ONLY SWSS UPG 3YR DISTI SO=, CSC IA CLD 3....
Name: description, dtype: object
************************************************************************************************************************
TOP CATEGORY FOR PRICE 1231
************************************************************************************************************************
TOP CATEGORY FOR QUANTITY 1539
************************************************************************************************************************


## Conclusion

### Products in stock

The top manufacturers for price and quantity were  HPE - SWITCHING and  DELL CSG SERVICE WARRANTIES. 

DELL CSG SERVICE WARRANTIES had the most profitable product, but this is still depepndent on the cost to produce the factor. Therefore this could still change.

The product for which there was an increase in price is 5YR PROSUPPORT PLUS, and the product for which there was a decrease in price is OPTIFIBER PRO QUAD OTDR V2 W/, INSPECTION & WIFI.

The top categories for price and quantity are 521 and 1221 respectively.

There are a few products which have 11 days of aging. It was also found that 43% of all the products (in stock) are aging (based on any product held for over 7 days).

### Products Not in Stock

The top manufacturers for price and quantity were  ALTNET and  MS-OPEN VALUE SUBSCRIPTION, respectively.

Based on the table created, 22MILES  had the most profitable product, however, profitability was calculated by multiplying the quantity by the price and since if the product is not in stock, the quantity is 0, all products have a profitability of 0.

The product for which there was an increase in price is CCW ONLY SWSS UPG 3YR DISTI SO=, CSC IA CLD 3, and the product for which there was a decrease in price is CCW ONLY SWSS UPG 3YR DISTI SO=, CSC IA CLD 3.

The top categories for price and quantity are 1231 and 1539 respectively.